x/ibc/04-channel/keeper/grpc_query.go

```diff
@@ -337,16 +337,58 @@ func (q Keeper) UnrelayedPackets(c context.Context, req *types.QueryUnrelayedPac
 			return nil, status.Errorf(codes.InvalidArgument, "packet sequence %d cannot be 0", i)
 		}
 
-		// if req.Acknowledgements is true append sequences with an existing acknowledgement
-		// otherwise append sequences without an existing acknowledgement.
-		if _, found := q.GetPacketAcknowledgement(ctx, req.PortId, req.ChannelId, seq); found == req.Acknowledgements {
+		// if acknowledgement exists on the receiving chain, then packet has already been received
+		if _, found := q.GetPacketAcknowledgement(ctx, req.PortId, req.ChannelId, seq); !found {
 			unrelayedSequences = append(unrelayedSequences, seq)
 		}
 
 	}
 
 	selfHeight := clienttypes.GetSelfHeight(ctx)
-	return &types.QueryUnrelayedPacketsResponse{
+	return &types.QueryUnreceivedPacketsResponse{
+		Sequences: unrelayedSequences,
+		Height:    selfHeight,
+	}, nil
+}
+
+// UnreceivedAcks implements the Query/UnreceivedAcks gRPC method. Given
+// a list of counterparty packet commitments, the querier checks if the packet
+// has already been received by checking if an acknowledgement exists on this
+// chain for the packet sequence. All packets that haven't been received yet
+// are returned in the response
+//
+// NOTE: The querier makes the assumption that the provided list of packet
+// commitments is correct and will not function properly if the list
+// is not up to date. Ideally the query height should equal the latest height
+// on the counterparty's client which represents this chain.
+func (q Keeper) UnreceivedAcks(c context.Context, req *types.QueryUnreceivedAcksRequest) (*types.QueryUnreceivedAcksResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	if err := validategRPCRequest(req.PortId, req.ChannelId); err != nil {
+		return nil, err
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+
+	var unrelayedSequences = []uint64{}
+
+	for i, seq := range req.PacketAckSequences {
+		if seq == 0 {
+			return nil, status.Errorf(codes.InvalidArgument, "packet sequence %d cannot be 0", i)
+		}
+
+		// if packet commitment still exists on the original sending chain, then packet ack has not been received
+		// since processing the ack will delete the packet commitment
+		if commitment := q.GetPacketCommitment(ctx, req.PortId, req.ChannelId, seq); len(commitment) != 0 {
```

Comment:
shouldn't this be querying if the acknowledgement exists? As per godoc
```

// Given a list of counterparty packet commitments, the querier checks if the packet
// has already been received by checking if an acknowledgement exists on this
// chain for the packet sequence.
```
I'm assuming the implemented flow here is trying to take in a list of acknowledgement sequences and check if the packet commitment exists to see the ack needs to be relayed? This seems inefficient. How does a relayer determine which acknowledgement sequences to send? Unlike packet commitments, acknowledgements live forever so a relayer would have to already know which acknowledgements have not been acknowledged on the original chain, making this function useless. 

The previous flow is to use the receiving chain as the indicator. A packet has 3 life cycles:
- sent but not received
- received but not ack'd
- ack'd and deleted

The sender chain indicates if the packet is in the first two stages (querying packet commitments) and the receiving chain indicates which of the first two stages it is in (ack doesn't exist, stage 1. ack does exist, stage 2).

Reply:
I assume the line of thinking that lead to this function is for it to be used in conjunction with the `QueryUnreceivedPackets`. This issue here arises with the introduction of async acknowledgements. Even if you relay unreceived packets, the acknowledgement may not be written for some time. You could keep a cache of those sequences, but you'd still need to query the receiving chain to find out if the acknowledgements have been written (verifying that your unreceived packet sequences has now become unreceived acknowledgements). You see what I am saying?

Regardless, the relayer **should not** rely on these functions beyond the short term. As per recent decisions in the IBC call, acknowledgements do not have to be written for proper functioning channels. This means packet commitments could live forever and relayers could get stuck trying to determine if an ack needs to be relayed. It should primarily rely on events

Reply:
> It seems the implemented flow here is trying to take in a list of acknowledgement sequences and check if the packet commitment exists to see the ack needs to be relayed? How does a relayer determine which acknowledgement sequences to send? Unlike packet commitments, acknowledgements live forever so a relayer would have to already know which acknowledgements have not been acknowledged on the original chain, which is the purpose of this function.

Ahh fair point. My thinking was that the relayer gives the entire list of acknowledgements to this function but that may be too large and inefficient in a long lived channel.

The original intention was to send packet commitments that still exist on the sending chain to the receiving chain and see if the ack exists? I can reimplement that, tho I still think it's better to make it a separate RPC call. Unfortunately this doesn't have a symmetry with the `QueryUnreceivedPackets` call which makes it non-intuitive.

> This issue here arises with the introduction of async acknowledgements. Even if you relay unreceived packets, the acknowledgement may not be written for some time. You could keep a cache of those sequences, but you'd still need to query the receiving chain to find out if the acknowledgements have been written (verifying that your unreceived packet sequences has now become unreceived acknowledgements). You see what I am saying?

Yes its possible even with the original implementation that once async acknowledgements are introduced that packet commitments exist for a long time before the ack gets written. I don't think this is a huge deal and hard to avoid. I think these functions should probably be used on relayer startup and once all packets in the queue are cleared, relayer should primarily rely on events